gateway-test/routes/oauth/createToken.js

```diff
@@ -0,0 +1,133 @@
+ */
+
+/**
+ * Create a New Token
+ * @route POST /oauth/token
+ * @param {TokenRequest.model} TokenRequest.body.required - the new token request
+ * @group users - Create a new Session Token for a user
+ * @returns {UserSession.model} 200 - An Object containing the user's data
+ * @returns {ErrorResponse.model}  default - OAuthError - Variable depending on request.
+ * @headers {string} 200.X-Expires-After -  date in UTC when token expires
+ */
+module.exports = (req, res) => {
+  if (req.body.hasOwnProperty('grant_type') === false) {
+    throw OAuthError.makeInvalidRequestError('grant_type')
+  }
+
+  if (req.body.grant_type === 'password') {
+    return handlePasswordGrant(req, res)
+  }
+
+  if (req.body.grant_type === 'refresh_token') {
+    return handleRefreshToken(req, res)
+  }
+
+  throw OAuthError.makeInvalidGrantError()
+}
+
+function handlePasswordGrant (req, res) {
+  switch (true) {
+    case req.body.hasOwnProperty('client_id') === false:
+      throw OAuthError.makeInvalidRequestError('client_id')
+    case clients.includes(client =>
+      client.clientId === req.body.client_id &&
+      client.clientSecret === req.body.client_secret
+    ):
+      throw OAuthError.makeInvalidClientError()
+    case req.body.hasOwnProperty('username') === false:
+      throw OAuthError.makeInvalidRequestError('username')
+    case req.body.hasOwnProperty('password') === false:
+      throw OAuthError.makeInvalidRequestError('password')
+  }
+
+  const user = users.find(obj => obj.email === req.body.username)
+
+  if (!user) {
+    throw OAuthError.makeInvalidCredentialsError()
+  }
+
+  if (user.password !== req.body.password) {
+    throw OAuthError.makeInvalidCredentialsError()
+  }
+
+  sendToken(res, user.id)
+}
+
+function handleRefreshToken (req, res) {
+  switch (true) {
+    case req.body.hasOwnProperty('client_id') === false:
+      throw OAuthError.makeInvalidRequestError('client_id')
+    case req.body.client_id != 'uniqueID':
+      throw OAuthError.makeInvalidClientError()
+    case req.body.hasOwnProperty('client_secret') === false:
+    case req.body.client_secret !== 'uniqueSecret':
+      throw OAuthError.makeInvalidClientError()
+    case req.body.hasOwnProperty('refresh_token') === false:
+      throw OAuthError.makeInvalidRequestError('refresh_token')
+  }
+
+  let decodedToken
+
+  try {
+    decodedToken = jwt.verify(req.body.refresh_token, 'ossp')
+  } catch (err) {
+    throw OAuthError.makeInvalidRefreshTokenError()
+  }
+
+  sendToken(res, decodedToken.sub)
+}
+
+function sendToken (res, subject) {
+  const options = {
+    subject
+  }
+
+  const UserSession = {
+    "access_token": jwt.sign({}, 'ossp', Object.assign(options, { expiresIn: '2 hours' })),
+    "token_type": "bearer",
+    "expires_in": 60 * 60 * 24,
+    "refresh_token": jwt.sign({}, 'ossp', Object.assign(options, { expiresIn: '2 days' }))
+  }
+
+  //Save this user to the database
+  axios.post('127.0.0.1:8080/user', {
```

Comment:
@SamNormcoreWayne  this is an example of what we should do when we need to call another service. This hits the database (port 8080) route /user, and we save the user's session into the db. If the database returns an error, we propagate that back to the user in the `.catch` block. Make sense?